### PR TITLE
Add IOST 

### PR DESCRIPTION
**Pull request by lispc**

### CODE REVIEW

**Comment by vikmeup** [comment]
@lispc can you rebase against the master? we don't commit generated code anymore

**Comment by hewigovens** [comment]
could you please keep code format in `TWCoinTypeConfiguration.cpp`? too many unrelated changes

**Comment by hewigovens** [comment]
this file is empty `src/IOST/Address.cpp`

File: swift/Tests/CoinAddressDerivationTests.swift [hunk 1]
```diff
@@ -79,6 +79,8 @@ class CoinAddressDerivationTests: XCTestCase {
                     XCTAssertEqual("0x0d1fa20c218Fec2f2C55d52aB267940485fa5DA4", address)
                 case .groestlcoin:
                     XCTAssertEqual("grs1qexwmshts5pdpeqglkl39zyl6693tmfwp0cue4j", address)
+                case .iost:
+                    XCTAssertTrue(true) // derived address is not available in IOST
```
Comment (hewigovens) [inline swift/Tests/CoinAddressDerivationTests.swift]: please return base58 representation

File: tests/IOST/SignerTests.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    std::string memo = "";
+
+    nlohmann::json j;
+    j.push_back(token);
+    j.push_back(src);
+    j.push_back(dst);
+    j.push_back(amount);
+    j.push_back(memo);
+    std::string data = j.dump();
+    ASSERT_EQ(data, std::string("[\"iost\",\"myname\",\"admin\",\"10\",\"\"]"));
```
Comment (hewigovens) [inline tests/IOST/SignerTests.cpp]: I have seen same json manipulation in kotlin/swift/cpp, could you please make it easier for coin transfer?

File: include/TrustWalletCore/TWIOSTAddress.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#include "TWData.h"
+#include "TWString.h"
+
+TW_EXTERN_C_BEGIN
+
+struct TWPublicKey;
+
+/// Represents an IOST address.
+TW_EXPORT_CLASS
+struct TWIOSTAddress;
```
Comment (hewigovens) [inline include/TrustWalletCore/TWIOSTAddress.h]:
```suggestion
struct TWIOSTPublickey;
```

File: src/Coin.cpp [hunk 4]
```diff
@@ -62,7 +63,8 @@ bool TW::validateAddress(TWCoinType coin, const std::string& string) {
 
     case TWCoinTypeGroestlcoin:
         return Bitcoin::Bech32Address::isValid(string, HRP_GROESTLCOIN) ||
-               Groestlcoin::Address::isValid(string, {TWP2PKHPrefixGroestlcoin, TWP2SHPrefixGroestlcoin});
+               Groestlcoin::Address::isValid(string,
```
Comment (hewigovens) [inline src/Coin.cpp]: please revert unrelated coins changes

File: src/Coin.cpp [hunk 5]
```diff
@@ -328,11 +337,8 @@ DerivationPath TW::derivationPath(TWCoinType coin) {
 
     case TWCoinTypeStellar:
     case TWCoinTypeKIN:
-        return DerivationPath{
-            DerivationPathIndex(purpose(coin), true),
-            DerivationPathIndex(coin, true),
-            DerivationPathIndex(0, true)
-        };
+        return DerivationPath{DerivationPathIndex(purpose(coin), true),
```
Comment (hewigovens) [inline src/Coin.cpp]: please revert unrelated coins changes

File: src/Coin.cpp [hunk 6]
```diff
@@ -421,6 +428,9 @@ std::string TW::deriveAddress(TWCoinType coin, const PublicKey& publicKey) {
         return Wanchain::Address(publicKey).string();
     case TWCoinTypeICON:
         return Icon::Address(publicKey, TWIconAddressTypeAddress).string();
+    case TWCoinTypeIOST:
+        // public key cannot derive address in IOST
+        return "";
```
Comment (hewigovens) [inline src/Coin.cpp]: please return base58 public key here

File: src/IOST/Address.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#include <stdint.h>
+#include <string>
+
+namespace TW {
+namespace IOST {
+
+class Address {
+  public:
+    template <typename T>
+    static bool isValid(const T& data) {
```
Comment (hewigovens) [inline src/IOST/Address.h]: This is for account name? then we should rename `Address` to `Account`

File: src/IOST/Signer.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+#include <iostream>
+
+using namespace TW;
+using namespace TW::IOST;
+
+class IOSTEncoder {
+  public:
+    IOSTEncoder() {}
+    void WriteByte(uint8_t b) { buffer << b; }
```
Comment (hewigovens) [inline src/IOST/Signer.cpp]: just curious we have `src/BinaryCoding.h`, could we reuse them?

File: src/interface/TWCoinTypeConfiguration.cpp [hunk 9]
```diff
@@ -174,6 +177,7 @@ const char *explorerURLForCoinType(enum TWCoinType type) {
     case TWCoinTypeGo: return "https://explorer.gochain.io";
     case TWCoinTypeGroestlcoin: return "https://blockbook.groestlcoin.org";
     case TWCoinTypeICON: return "https://tracker.icon.foundation";
+    case TWCoinTypeIOST: return "https://www.iostabc.com";
```
Comment (hewigovens) [inline src/interface/TWCoinTypeConfiguration.cpp]: Is this official?

File: src/IOST/Signer.cpp (hunk 8 above)
Comment (lispc) [reply src/IOST/Signer.cpp]: the encoding method is related to signature, so we have to implement it

File: include/TrustWalletCore/TWIOSTAddress.h (hunk 3 above)
Comment (lispc) [reply include/TrustWalletCore/TWIOSTAddress.h]: changed it to account

File: tests/IOST/SignerTests.cpp (hunk 2 above)
Comment (lispc) [reply tests/IOST/SignerTests.cpp]: done

File: src/IOST/Signer.cpp (hunk 8 above)
Comment (hewigovens) [reply src/IOST/Signer.cpp]: please take a look at `BinaryCoding::encode16BE`

File: swift/Sources/Addresses/CoinType+Address.swift [hunk 10]
```diff
@@ -63,6 +63,8 @@ public extension CoinType {
             } else {
                 return GroestlcoinAddress(string: string)
             }
+        case .iost:
+            return IOSTAddress(string: string)
```
Comment (hewigovens) [inline swift/Sources/Addresses/CoinType+Address.swift]: should be `IOSTAccount`?

File: include/TrustWalletCore/TWIOSTSigner.h [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+#pragma once
+
+#include "TWBase.h"
+#include "TWData.h"
+#include "TWIOSTProto.h"
+
+TW_EXTERN_C_BEGIN
+
+/// Helper class to sign Wanchain transactions.
```
Comment (hewigovens) [inline include/TrustWalletCore/TWIOSTSigner.h]:
```suggestion
/// Helper class to sign IOST transactions.
```

File: src/IOST/Account.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+#pragma once
+
+#include "../PublicKey.h"
+#include "../proto/IOST.pb.h"
+#include "../Base58.h"
+
+#include <stdint.h>
+#include <string>
+
+namespace TW {
```
Comment (hewigovens) [inline src/IOST/Account.h]:
```suggestion
namespace TW::IOST {
```

File: src/IOST/Account.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    static std::string encodePubKey(const PublicKey& publicKey) {
+        return Base58::bitcoin.encode(publicKey.bytes);
+    } 
+    Account(const std::string& name): name(name) {}
+    Account(const PublicKey& publicKey) {}
+    Account(const Proto::AccountInfo& account): name(account.name()), activeKey(account.active_key()), ownerKey(account.owner_key()) {}
+    std::string string() const { return name; }
+    std::string name;
+    std::string activeKey;
+    std::string ownerKey;
```
Comment (hewigovens) [inline src/IOST/Account.h]: This is hex string or base58 string? if it's hex string, maybe we can use bytes or `PublicKey` directly?

File: src/IOST/Account.cpp [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+using namespace TW;
+using namespace TW::IOST;
+
+bool Account::isValid(const std::string& s) {
+    if (s.size() < 5 || s.size() > 11) { 
+        return false; 
+    }
+    for (char ch : s) {
+        if (!((ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || ch == '_')) {
```
Comment (alejandro-isaza) [inline src/IOST/Account.cpp]: I find this easier to follow:
```suggestion
        if ((ch < 'a' || ch > 'z') && (ch < '0' || ch > '9') && ch != '_') {
```

File: src/IOST/Account.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+#include <stdint.h>
+#include <string>
+
+namespace TW::IOST {
+
+class Account {
+  public:
+    static bool isValid(const std::string& name);
+    static std::string encodePubKey(const PublicKey& publicKey);
+    Account(const std::string& name): name(name) {}
```
Comment (alejandro-isaza) [inline src/IOST/Account.h]: See clang-tidy warnings: https://dev.azure.com/TrustWallet/Trust%20Wallet%20Core/_build/results?buildId=2388
```suggestion
    Account(std::string name): name(std::move(name)) {}
```

File: src/IOST/Signer.cpp [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+#include <iostream>
+
+#include <nlohmann/json.hpp>
+
+using namespace TW;
+using namespace TW::IOST;
+
+class IOSTEncoder {
+  public:
+    IOSTEncoder() {}
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]: /wallet-core/src/IOST/Signer.cpp:16:5: warning: use '= default' to define a trivial default constructor [modernize-use-equals-default]

File: src/IOST/Signer.cpp [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    void WriteByte(uint8_t b) { buffer << b; }
+    void WriteInt32(uint32_t i) {
+        uint32_t r = boost::endian::endian_reverse(i);
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]: This will depend on the machine architecture. Please use the functions in `BinaryCoding.h`

File: src/IOST/Signer.cpp [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        buffer << std::string(reinterpret_cast<const char*>(&r), sizeof(r));
+    }
+
+    void WriteInt64(uint64_t i) {
+        uint64_t r = boost::endian::endian_reverse(i);
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]: Same here

File: src/IOST/Signer.cpp [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    se.WriteString("");
+
+    std::vector<std::string> svec;
+    for (auto item : t.signers()) {
+        svec.push_back(item);
+    }
+    se.WriteStringSlice(svec);
+
+    se.WriteInt32(t.actions_size());
+    for (auto a : t.actions()) {
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]:
```suggestion
    for (auto& a : t.actions()) {
```

File: src/IOST/Signer.cpp [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+    se.WriteInt64(t.time());
+    se.WriteInt64(t.expiration());
+    se.WriteInt64(int64_t(t.gas_ratio() * 100));
+    se.WriteInt64(int64_t(t.gas_limit() * 100));
+    se.WriteInt64(t.delay());
+    se.WriteInt32(int32_t(t.chain_id()));
+    se.WriteString("");
+
+    std::vector<std::string> svec;
+    for (auto item : t.signers()) {
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]:
```suggestion
    for (auto& item : t.signers()) {
```

File: src/IOST/Signer.cpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+    for (auto a : t.actions()) {
+        IOSTEncoder s;
+        s.WriteString(a.contract());
+        s.WriteString(a.action_name());
+        s.WriteString(a.data());
+        se.WriteString(s.AsString());
+    }
+
+    se.WriteInt32(t.amount_limit_size());
+    for (auto a : t.amount_limit()) {
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]:
```suggestion
    for (auto& a : t.amount_limit()) {
```

File: src/IOST/Signer.cpp [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+        IOSTEncoder s;
+        s.WriteString(a.token());
+        s.WriteString(a.value());
+        se.WriteString(s.AsString());
+    }
+
+    se.WriteInt32(t.signatures_size());
+    for (auto sig : t.signatures()) {
```
Comment (alejandro-isaza) [inline src/IOST/Signer.cpp]:
```suggestion
    for (auto& sig : t.signatures()) {
```

File: src/IOST/Signer.h [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include "../proto/IOST.pb.h"
+#include "../PrivateKey.h"
+
+namespace TW {
+namespace IOST {
```
Comment (alejandro-isaza) [inline src/IOST/Signer.h]: `namespace TW::IOST`

File: src/IOST/Signer.h [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+class Signer {
+public:
+    Proto::SigningOutput sign(const Proto::SigningInput& input) const noexcept;
+};
+
+}} // namespace
```
Comment (alejandro-isaza) [inline src/IOST/Signer.h]:
```suggestion
}} // namespace TW::IOST
```

File: src/IOST/Account.cpp (hunk 14 above)
Comment (lispc) [reply src/IOST/Account.cpp]: I think the original version is more friendly for readers?

**Comment by lispc** [comment]
@alejandro-isaza updated the codes here
